### PR TITLE
Conversion check for cycle introduction and transitivity check

### PR DESCRIPTION
**Pull request by insafuhrmann**
Changes in conversion.ts:
Eliminated TODOs by implementing graph algorithm to be used in check for introduction of cycles through implicit conversion and for transitive conversion checks.

Introduces error in markAsConvertibleSigne in case an implicit conversion has introduced a cycle in the type graph.

### CODE REVIEW

File: packages/typir/src/features/conversion.ts [hunk 1]
```diff
@@ -136,20 +136,16 @@ export class DefaultTypeConversion implements TypeConversion {
...
         if (mode === 'IMPLICIT_EXPLICIT') {
-            this.checkForCyclesLogic(mode);
-        } else {
-            // all other modes allow cycles
+            /* check that the new edges did not introduce cycles
+             * if it did, the from node will be reachable via a cycle path
+             */
+            const hasIntroducedCycle = this.existsEdgePath(from, from, mode);
+            if (hasIntroducedCycle) {
+                throw new Error('Adding the conversion from ' + from.identifier + ' to ' + to.identifier + ' has introduced a cycle ot the type graph.');
```
Comment (JohannesMeierSE) [inline packages/typir/src/features/conversion.ts:145]: - Maybe adding the `mode` as well would make the error message even more clear?
- typo: "ot"

File: packages/typir/src/features/conversion.ts [hunk 2]
```diff
@@ -180,11 +176,47 @@ export class DefaultTypeConversion implements TypeConversion {
         return 'NONE';
     }
 
-    protected isTransitivelyConvertable(_from: Type, _to: Type, _mode: ConversionModeForSpecification): boolean {
-        // TODO calculate transitive relationship
+    protected existsEdgePath(from: Type, to: Type, mode: ConversionModeForSpecification): boolean {
+        const visited: Set<Type> = new Set();
+        const stack: Type[] = [from];
+
+        while (stack.length > 0) {
```
Comment (JohannesMeierSE) [inline packages/typir/src/features/conversion.ts:183]: Thanks for the iterative implementation!

**Review by JohannesMeierSE — APPROVED** [review]
Thanks for this implementation! It looks very nice!

I found only a typo. After fixing, feel free to merge this PR on `main` yourself 🙂

File: packages/typir/src/features/conversion.ts (hunk 2 above)
Comment (insafuhrmann) [reply packages/typir/src/features/conversion.ts:183]: I hope it helps with performance.

File: packages/typir/src/features/conversion.ts (hunk 1 above)
Comment (insafuhrmann) [reply packages/typir/src/features/conversion.ts]: Ah, good catch! Thanks, and thanks for reviewing!

**Comment by insafuhrmann** [comment]
@JohannesMeierSE Thanks for reviewing and approval. The typo is fixed now.

**Pull request merged**